### PR TITLE
#8535: Additional improvement on the main bundle

### PR DESCRIPTION
**Pull request by allyoucanmap**
## Description

This PR includes improvement related to:

- [x] Importing components libraries only when they are mounted in the UI using lazy + suspense. Here the libraries:
  - [x] react-codemirror
  - [x] react-draft-wysiwyg
  - [x] pdf.js
  - [x] react-quill
  - [x] react-data-grid
  - [x] pdf.js
  - [x] vis-timeline-graph2d.min
- [x] Completely removed `css-tree` in favor of static json configuration
- [x] Importing only the woff2 font for icons in the css file
- [x] made the browser valid check script async in the html page
- [x] remove leaflet cdn script and css from html pages in favor of dynamic loading

This PR introduces also `@testing-library/react` to simplify test of lazy react components

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x", remove the others)
 - [x] Other... Please describe: Improvement on dependencies loading

## Issue

**What is the current behavior?**

#8535

**What is the new behavior?**

## Breaking change
**Does this PR introduce a breaking change?** (check one with "x", remove the other)
 - [x] Yes, and I documented them in migration notes

### CODE REVIEW

**Review by offtherailz — CHANGES_REQUESTED** [review]
I can not fully review.
- [x] I can suggest to add the "Migration guidelines" for html files.

Apart from this PR we may do a raw test to see how the score improves if we setup caching for every js/css file. If it makes the difference, we should make the main
- bundles 
-  css 
-  extensions main files should be excluded for the moment(index.js has not hash) 
to have the hash in the name, so caching can be applied easely and updates will load new files.

File: docs/developer-guide/mapstore-migration-guide.md [hunk 1]
```diff
@@ -22,6 +22,41 @@ This is a list of things to check if you want to update from a previous version
...
+- <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/leaflet/1.3.1/leaflet.css" />
+- <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/leaflet.draw/1.0.2/leaflet.draw.css" />
+<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" />
+<link rel="shortcut icon" type="image/png" href="https://cdn.jslibs.mapstore2.geo-solutions.it/leaflet/favicon.ico" />
+<!--script src="https://maps.google.com/maps/api/js?v=3"></script-->
+- <script src="https://cdnjs.cloudflare.com/ajax/libs/leaflet/1.3.1/leaflet.js"></script>
+- <script src="https://cdnjs.cloudflare.com/ajax/libs/leaflet.draw/1.0.2/leaflet.draw.js"></script>
+```
+
+We also made asynchronous the script to detect valid browser this should slightly improve the initial requests time, you can updated the script in your project as follow:
```
Comment (offtherailz) [inline docs/developer-guide/mapstore-migration-guide.md:39]:
```suggestion
We also made asynchronous the script to detect valid browser. This should slightly improve the initial requests time.
 You can updated the script in your project as following:

```

**Review by offtherailz — CHANGES_REQUESTED** [review]
LGTM for my inspection. Just fix the suggested text.

File: web/client/components/background/BackgroundSelector.jsx [hunk 2]
```diff
@@ -6,17 +6,19 @@
...
 import PreviewButton from './PreviewButton';
 import PreviewList from './PreviewList';
 import PreviewIcon from './PreviewIcon';
 import ToolbarButton from '../misc/toolbar/ToolbarButton';
-import BackgroundDialog from './BackgroundDialog';
+
 import ConfirmDialog from '../misc/ConfirmDialog';
 import PropTypes from 'prop-types';
+import withSuspense from '../misc/withSuspense';
+const BackgroundDialog = withSuspense()(lazy(() => import('./BackgroundDialog')));
```
Comment (MV88) [inline web/client/components/background/BackgroundSelector.jsx:21]: should we document this new approach somewhere? maybe in the training?

File: web/client/components/misc/withSuspense.jsx [hunk 3]
```diff
@@ -0,0 +1,20 @@
+/*
+ * Copyright 2022, GeoSolutions Sas.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root dir
+ ectory of this source tree.
+ */
```
Comment (MV88) [inline web/client/components/misc/withSuspense.jsx:8]:
```suggestion
 * LICENSE file in the root directory of this source tree. */
```

File: web/client/components/misc/withSuspense.jsx [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+import React, { Suspense, forwardRef } from 'react';
+
+const withSuspense = (activeFunc, { fallback = null } = {}) => (Component) => forwardRef((props, ref) => {
```
Comment (MV88) [inline web/client/components/misc/withSuspense.jsx:12]: can you add jsdoc with an example?

File: web/client/components/styleeditor/Editor.jsx [hunk 5]
```diff
@@ -236,7 +225,7 @@ class Editor extends React.Component {
                                 this.props.onChange(code);
                             }
                         }, this.props.waitTime);
-                        CM.extendMode(this.props.mode, { hintProperties: this.props.hintProperties });
+                        cm.extendMode(this.props.mode, { hintProperties: this.props.hintProperties });
```
Comment (MV88) [inline web/client/components/styleeditor/Editor.jsx:228]: this.cm is better, right?
```suggestion
                        this.cm.extendMode(this.props.mode, { hintProperties: this.props.hintProperties });
```

File: web/client/libs/codemirror/react-codemirror2.js [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+import cm from 'codemirror/lib/codemirror';
+
+import geoCssMode from './mode/geocss';
+import geoCssHint from './hint/geocss';
+
+geoCssMode(cm);
+geoCssHint(cm);
+
+function CodeMirror({ editorDidMount = () => {}, ...props}, ref) {
```
Comment (MV88) [inline web/client/libs/codemirror/react-codemirror2.js:26]: do we have to document why this has been moved here?

File: web/client/reducers/mapInfo.js [hunk 7]
```diff
@@ -42,6 +42,8 @@ import { findIndex, isUndefined, isEmpty } from 'lodash';
 import { MAP_TYPE_CHANGED } from './../actions/maptype';
 
 import { getValidator } from '../utils/MapInfoUtils';
+import buffer from 'turf-buffer';
+import intersect from 'turf-intersect';
```
Comment (MV88) [inline web/client/reducers/mapInfo.js:46]: can these be imported among other libs?

**Review by MV88 — CHANGES_REQUESTED** [review]
Just a few comments here and there, mostly related to the documentation

anyway the app seems faster

File: web/client/components/background/BackgroundSelector.jsx (hunk 2 above)
Comment (allyoucanmap) [reply web/client/components/background/BackgroundSelector.jsx:21]: I agree that we should add a documentation to explain how to create a plugin taking into account the new store api and the lazy approach.
In this issue there is an item on the list to cover this request https://github.com/geosolutions-it/MapStore2/issues/6735#issuecomment-1233077463

Comment (MV88) [reply web/client/components/background/BackgroundSelector.jsx:21]: ok good

**Review by MV88 — APPROVED** [review]

**Review by offtherailz — APPROVED** [review]
Thank you for the changes on doc. :+1:

**Comment by allyoucanmap** [comment]
@tdipisa @ElenaGallo this PR can be tested on DEV, we should compare the initial loading with QA and 2022.01.00 instances. As soon as everything is tested we can backport to stable branch 2022.02.xx

**Comment by MV88** [comment]
@allyoucanmap can you start preparing the backport?

**Pull request merged**